imageio/core/imopen.py

```diff
@@ -0,0 +1,477 @@
+    """ Open a URI and return a plugin instance that can read its content.
+
+    Imopen is the plugin manager of imopen. It acts as a factory for Plugin
+    objects implementing the v3.0 API. It is the core of the v3.0 API and the
+    functional API provides convenient wrappers around this object.
+
+    Note for library maintainers: If you call imopen from inside the library,
+    you will first need to first create an instance, i.e.
+    ``imopen()(uri, ...)``. Note the double parentheses.
+
+    """
+    _known_plugins = dict()
+    _legacy_format_manager = FormatManager()
+
+    def __call__(self, uri, *args, plugin=None, legacy_api=True, **kwargs):
+        """ Instantiate a plugin capable of opening URI
+
+        Parameters
+        ----------
+        uri : {str, pathlib.Path, bytes, file}
+            The resource to load the image from, e.g. a filename, pathlib.Path,
+            http address or file object, see the docs for more info.
+        *args :
+            Additional positional arguments will be passed to the plugin instance
+            upon object creation.
+        plugin : {str, None}
+            The plugin to be used. If None, performs a search for a matching plugin.
+        legacy_api : {bool}
+            The API to be used. If true use the v2.9 API, otherwise use the new
+            v3.0 API.
+        **kwargs :
+            Additional keyword arguments will be passed to the plugin instance.
+        """
+
+        plugin_instance = None
+
+        if legacy_api:
+            kwargs["plugin"] = plugin
+            return LegacyPlugin(
+                uri,
+                self._legacy_format_manager,
+                *args,
+                **kwargs
+            )
+
+        if plugin is not None:
+            try:
+                plugin_instance = self._known_plugins[plugin]
+            except KeyError:
+                raise ValueError(
+                    f"'{plugin}' is not a registered plugin name.")
+
+        else:
+            for candidate_plugin in self._known_plugins.values():
+                if (candidate_plugin.can_open(uri)):
+                    plugin_instance = candidate_plugin
+                    break
+            else:
+                raise IOError(f"No registered plugin can read {uri}")
+
+        return plugin_instance(uri, *args, **kwargs)
+
+    @classmethod
+    def register_plugin(cls, plugin_name, plugin_class):
+        """ Register a new plugin to be used when opening URIs.
+
+        Parameters
+        ----------
+        plugin_name : str
+            The name of the plugin to be registered. If the name already exists
+            it will be overwritten.
+        plugin_class : callable
+            A callable that returns an instance of a plugin that conforms
+            to the v3.0 API.
+        """
+
+        cls._known_plugins[plugin_name] = plugin_class
+
+
+class Plugin(object):
+    def __init__(self, uri):
+        """ Instantiate a new Legacy Plugin
```

Comment:
FWIW, I think plugins should follow a specification, but should not need to inherit from a specific class. This e.g. makes it possible to provide plugins in other packages without these packages needing to rely on imageio. It also forces us to keep the spec for plugins as simple as possible.

Reply:
Also, why does it say this a Legacy Plugin?

Reply:
I agree; inheriting is not required (as long as the plugin adheres to the interface). The reason I introduced a generic Plugin class is purely for my convenience of not having to write duplicate code for the legacy plugin and the pillow plugin. I can also remove it and duplicate code where necessary. Which option do you prefer?

The `Legacy Plugin` is a copy-paste error when I copied the docstrings. I'll fix it before we merge the PR.

Reply:
If it helps prevent duplicating code, we should do that. But we should document its purpose. Maybe we can even implement a `PluginInterface` class that servers solely to define what a plugin should look like, and an `ImageioPlugin` for use inside imageio. Anyway, it should be clear for people reading this code that one does not need to inherit from it. Perhaps a comment dufficces :)

Reply:
On second thought, the amount of duplication we saved was less than the boilerplate added by the class. I've removed it :)